src/data/finsupp.lean

```diff
@@ -1825,3 +1825,124 @@ lemma finite_lt_nat (n : σ →₀ ℕ) : set.finite {m | m < n} :=
 (finite_le_nat n).subset $ λ m, le_of_lt
 
 end finsupp
+
+section finsupp_lattice
+variable [has_zero β]
+
+--these two need a new home
+@[simp]
+lemma nat.inf_zero_iff {m n : ℕ} : m ⊓ n = 0 ↔ m = 0 ∨ n = 0 :=
+begin
+  split, swap, intro h, cases h; rw h; simp,
+  unfold has_inf.inf, unfold semilattice_inf.inf, unfold lattice.inf, unfold min,
+  intro hmin, by_cases m = 0, left, apply h,
+  rw if_neg _ at hmin, right, apply hmin, contrapose hmin,
+  simp only [not_lt, not_le] at hmin, rw if_pos hmin, apply h
+end
+
+lemma nat.zero_eq_bot : (0 : ℕ) = ⊥ := rfl
+
+namespace finsupp
+
+instance has_inf [semilattice_inf β] :
+  has_inf (α →₀ β) :=
+begin
+  refine ⟨_⟩, intros v1 v2,
+  refine ⟨(v1.support ∪ v2.support).filter (λ x, v1 x ⊓ v2 x ≠ 0),
+    (λ (a : α), (v1 a ⊓ v2 a)), _⟩,
+  intro a, simp only [finsupp.mem_support_iff, ne.def, finset.mem_union, finset.mem_filter],
+  apply and_iff_right_of_imp, contrapose, push_neg,
+  intro h, rw [h.left, h.right], apply inf_idem,
+end
+
+@[simp]
+lemma inf_apply [semilattice_inf β] {a : α} {f g : α →₀ β} : (f ⊓ g) a = f a ⊓ g a := rfl
+
+@[simp]
+lemma nat.support_inf {f g : α →₀ ℕ} : (f ⊓ g).support = f.support ∩ g.support :=
```

Comment:
Can this be generalised? What was the outcome of the Zulip discussion? I can imagine we want a similar thing for `nnreal` and/or `enat` at some point.

Reply:
I've generalized it somewhat, but I still have some weird assumptions. I'd like to take care of them by assuming [canonically_ordered_add_monoid] and [decidable_linear_order], but I'd end up with 3 competing definitions of <= and at probably two competing definitions of 0.